juju/charm.py

```diff
@@ -94,11 +97,55 @@ def __init__(self, framework, key, metadata, charm_dir):
             self.on.define_event(f'{relation_name}_relation_departed', RelationDepartedEvent)
             self.on.define_event(f'{relation_name}_relation_broken', RelationBrokenEvent)
 
-        for storage_name in metadata.storage:
+        for storage_name in self.metadata.storage:
             storage_name = storage_name.replace('-', '_')
             self.on.define_event(f'{storage_name}_storage_attached', StorageAttachedEvent)
             self.on.define_event(f'{storage_name}_storage_detaching', StorageDetachingEvent)
 
+    @property
+    def charm_env(self):
+        return self.framework.charm_env
+
+    @property
+    def metadata(self):
+        return self.charm_env.metadata
+
+    @property
+    def charm_dir(self):
+        return self.charm_env.charm_dir
+
+
+class CharmEnv:
+    def __init__(self, env=None, **overrides):
+        """Object containing info about the operating environment of the charm.
```

Comment:
If we give people an easy programmatic access to the _current_ environment via the framework, it means handling an event the first time vs. the second time may result in completely different behavior. Consider what would happen, for example, once we change this system to be fed via events that come from a socket, without the process dying.

This is a good topic for us to cover in the meeting today.